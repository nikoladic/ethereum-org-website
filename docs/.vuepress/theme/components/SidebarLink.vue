<script>
import {
  isActive,
  hashRE,
  groupHeaders,
  resolveHeaderTitle
} from '../utils/util'

export default {
  functional: true,

  props: ['item'],

  render(h, { parent: { $page, $site, $route }, props: { item } }) {
    // use custom active class matching logic
    // due to edge case of paths ending with / + hash
    const selfActive = isActive($route, item.path)
    // for sidebar: auto pages, a hash link should be active if one of its child
    // matches
    const active =
      item.type === 'auto'
        ? selfActive ||
          item.children.some(c =>
            isActive($route, item.basePath + '#' + c.slug)
          )
        : selfActive
    const link = renderLink(h, item.path, item.title || item.path, active)
    const configDepth =
      $page.frontmatter.sidebarDepth != null
        ? $page.frontmatter.sidebarDepth
        : $site.themeConfig.sidebarDepth
    const maxDepth = configDepth == null ? 1 : configDepth
    const displayAllHeaders = !!$site.themeConfig.displayAllHeaders
    if (item.type === 'auto') {
      return h('div', {}, [
        link,
        renderChildren(h, item.children, item.basePath, $route, maxDepth)
      ])
    } else if (
      (active || displayAllHeaders) &&
      item.headers &&
      !hashRE.test(item.path)
    ) {
      const children = groupHeaders(item.headers)
      return h('div', {}, [
        link,
        renderChildren(h, children, item.path, $route, maxDepth)
      ])
    } else {
      return link
    }
  }
}

function renderLink(h, to, text, active, isSubHeader = false) {
  return h(
    'a',
    {
      attrs: {
        href: to
      },
      class: {
        'tc-primary500 active': active,
        'bc-primary500': active && !isSubHeader,
        'tc-text100': !active,
        'sidebar-link br-25rem w-100 l8 tc-h-primary500 mt-0 mb-05 pr-025': true
      }
    },
    resolveHeaderTitle(text)
  )
}

function renderChildren(h, children, path, route, maxDepth, depth = 1) {
  if (!children || depth > maxDepth) return null
  return h(
    'ul',
    { class: 'pl-1' },
    children.map(c => {
      const active = isActive(route, path + '#' + c.slug)
      return h('li', { class: 'pl-1' }, [
        renderLink(h, path + '#' + c.slug, c.title, active, true),
        renderChildren(h, c.children, path, route, maxDepth, depth + 1)
      ])
    })
  )
}
</script>
