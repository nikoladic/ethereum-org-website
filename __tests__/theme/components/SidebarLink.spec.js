import { shallowMount } from '@vue/test-utils'
import SidebarLink from "@/theme/components/SidebarLink"


describe("As user i see a sidebar link", () => {
  test('Rendered without childs', () => {
    const wrapper = shallowMount(SidebarLink, {
      propsData: {
        item: {
          type: 'auto', path:'/whales', basePath: '/', slug: 'big', title: 'Whales', headers: ['whales'], children: []
        },
        collapsable: true,
        first: true,
        open: true
      },
      mocks: {
        $page: {frontmatter: {sidebarDepth: 1}},
        $site: { themeConfig: {displayAllHeaders: true}},
        $route: {'hash': '89452c3t7c0923479827'},
        $siteTitle: 'Click me!'
      }
    });

    expect(wrapper.element).toMatchSnapshot()
  });

  test('Rendered with childs', () => {
    const wrapper = shallowMount(SidebarLink, {
      propsData: {
        item: {
          type: 'auto', path:'/whales', basePath: '/', slug: 'big', title: 'Whales', headers: ['whales'], children: [
            { slug: 'big', title: 'Whales'}
          ]
        },
        collapsable: true,
        first: true,
        open: true
      },
      mocks: {
        $page: {frontmatter: {sidebarDepth: 1}},
        $site: { themeConfig: {displayAllHeaders: true}},
        $route: {'hash': '89452c3t7c0923479827'},
        $siteTitle: 'Click me!'
      }
    });

    expect(wrapper.element).toMatchSnapshot()
  });

  test('Rendered without display all headers', () => {
    const wrapper = shallowMount(SidebarLink, {
      propsData: {
        item: {
          type: 'auto', path:'/whales', basePath: '/', slug: 'big', title: 'Whales', headers: ['whales'], children: []
        },
        collapsable: true,
        first: true,
        open: true
      },
      mocks: {
        $page: {frontmatter: {sidebarDepth: 1}},
        $site: { themeConfig: {displayAllHeaders: false}},
        $route: {'hash': '89452c3t7c0923479827'},
        $siteTitle: 'Click me!'
      }
    });

    expect(wrapper.element).toMatchSnapshot()
  });
});
