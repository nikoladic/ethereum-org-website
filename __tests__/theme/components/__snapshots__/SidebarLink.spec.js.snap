// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`As user i see a sidebar link Rendered with childs 1`] = `<!---->`;

exports[`As user i see a sidebar link Rendered without childs 1`] = `<!---->`;

exports[`As user i see a sidebar link Rendered without display all headers 1`] = `<!---->`;
